internal/asm/arm64/impl.go

```diff
@@ -0,0 +1,2340 @@
+	OperandTypesSIMDByteToRegister             = OperandTypes{OperandTypeSIMDByte, OperandTypeRegister}
+	OperandTypesTwoSIMDBytesToSIMDByteRegister = OperandTypes{OperandTypeTwoSIMDBytes, OperandTypeSIMDByte}
+)
+
+// String implements fmt.Stringer
+func (o OperandTypes) String() string {
+	return fmt.Sprintf("from:%s,to:%s", o.src, o.dst)
+}
+
+// AssemblerImpl implements Assembler.
+type AssemblerImpl struct {
+	asm.BaseAssemblerImpl
+	Root, Current     *NodeImpl
+	Buf               *bytes.Buffer
+	temporaryRegister asm.Register
+	nodeCount         int
+	pool              constPool
+}
+
+// constPool holds 32-bit constants which are used by ldr(literal) instructions
+// emitted by memory access.
+type constPool struct {
+	// firstUseOffsetInBinary is the offset of the first ldr(literal) instruction
+	// which needs to access the const in this constPool.
+	firstUseOffsetInBinary *asm.NodeOffsetInBinary
+	consts                 []int32
+	// offsetFinalizedCallbacks holds the callbacks keyed on the constats.
+	// These callbacks are called when the offsets of the constants in the binary
+	// have been determined.
+	offsetFinalizedCallbacks map[int32][]func(offsetOfConstInBinary int)
+}
+
+func NewAssemblerImpl(temporaryRegister asm.Register) *AssemblerImpl {
+	return &AssemblerImpl{
+		Buf: bytes.NewBuffer(nil), temporaryRegister: temporaryRegister,
+		pool: constPool{offsetFinalizedCallbacks: map[int32][]func(int){}},
+	}
+}
+
+// newNode creates a new Node and appends it into the linked list.
+func (a *AssemblerImpl) newNode(instruction asm.Instruction, types OperandTypes) *NodeImpl {
+	n := &NodeImpl{
+		Instruction: instruction,
+		Next:        nil,
+		Types:       types,
+		JumpOrigins: map[*NodeImpl]struct{}{},
+	}
+
+	a.addNode(n)
+	return n
+}
+
+// addNode appends the new node into the linked list.
+func (a *AssemblerImpl) addNode(node *NodeImpl) {
+	a.nodeCount++
+
+	if a.Root == nil {
+		a.Root = node
+		a.Current = node
+	} else {
+		parent := a.Current
+		parent.Next = node
+		a.Current = node
+	}
+
+	for _, o := range a.SetBranchTargetOnNextNodes {
+		origin := o.(*NodeImpl)
+		origin.JumpTarget = node
+	}
+	a.SetBranchTargetOnNextNodes = nil
+}
+
+// Assemble implements asm.AssemblerBase
+func (a *AssemblerImpl) Assemble() ([]byte, error) {
+	// arm64 has 32-bit fixed length instructions,
+	// but note that some of nodes are encoded as multiple instructions,
+	// so the resulting binary might not be the size of count*8.
+	a.Buf.Grow(a.nodeCount * 8)
+
+	for n := a.Root; n != nil; n = n.Next {
+		n.OffsetInBinaryField = (uint64(a.Buf.Len()))
+		if err := a.EncodeNode(n); err != nil {
+			return nil, fmt.Errorf("%w: %v", err, n)
+		}
+		a.maybeFlushConstPool(n.Next == nil)
+	}
+
+	code := a.Bytes()
+	for _, cb := range a.OnGenerateCallbacks {
+		if err := cb(code); err != nil {
+			return nil, err
+		}
+	}
+	return code, nil
+}
+
+// maybeFlushConstPool maybe flushes the current constants in the constPool.
```

Comment:
maybeFlushConstPool flushes the constant pool if endOfBinary or a boundary condition was met.